Make Allocation::bytes private

197g opened the pull request:
Fixes #62931.

Direct immutable access to the bytes is still possible but redirected through the new method `raw_bytes_with_undef_and_ptr`, similar to `get_bytes_with_undef_and_ptr` but without requiring an interpretation context and not doing *any* relocation or bounds checks. The `size` of the allocation is stored separately which makes access as `Size` and `usize` more ergonomic.

cc: @RalfJung

197g committed:
Store allocation size, make bytes, undef_mask private

Direct access to the bytes was previously a problem (#62931) where
components would read their contents without properly checking
relocations and/or definedness.

Making bytes private instead of purely renaming them also helps in
allowing amendments to their allocation scheme (such as eliding
allocation for undef of constant regions).

197g committed:
Replace usage of alloc.bytes in interpret

There is now a dedicate `len` method which avoids the need to access the
bytes. Access the length as `Size` can also be done by a direct member.
The constructors guarantee that these representations are convertable.

Access which relies on the bytes, such as snapshot, can use direct raw
access by reference as it does not care about undef and relocations or
properly checks them seperately.

197g committed:
Move copy of undef_mask into allocation

This also means that the compressed representation chosen may be
optimized together with any changes to the undef_mask.

197g committed:
Expose encapsulated undef mask as immutable

197g committed:
Fix codegen with explicit allocation byte access

rust-highfive commented:
r? @cramertj

(rust_highfive has picked a reviewer for you, use r? to override)

Centril commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -16,15 +16,20 @@ use std::borrow::Cow;
+    /// Note that the bytes of a pointer represent the offset of the pointer. Bytes of undefined
+    /// allocations may be allocated lazily as bytes are initialized but either none or all bytes
+    /// are allocated.
+    pub(crate) bytes: Vec<u8>,
```
```suggestion
    crate bytes: Vec<u8>,
```

HeroicKatora replied:
Requires a customized implementation in `src/librustc/ich/impls_ty.rs` where destructuring assignment could no longer capture all fields. Would that be fine?

RalfJung replied:
`crate` and `pub(crate)` are equivalent. Why would anything about that impl change?

HeroicKatora replied:
Oh. Right. My bad, I'm just not used to `crate field` syntax.

Centril commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -16,15 +16,20 @@ use std::borrow::Cow;
-    /// Denotes undefined memory. Reading from undefined memory is forbidden in miri
-    pub undef_mask: UndefMask,
+    /// Denotes which part of this allocation is initialized
+    pub(crate) undef_mask: UndefMask,
```
```suggestion
    crate undef_mask: UndefMask,
```

HeroicKatora commented:
r? @RalfJung

HeroicKatora commented:
Changed `pub(crate)` to `crate` by amending the first commit.

RalfJung commented:
I edited your OP to say "Fixes: #issue"; that will make GitHub auto-close it when the PR gets merged.

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -16,15 +16,20 @@ use std::borrow::Cow;
+    /// Note that the bytes of a pointer represent the offset of the pointer. Bytes of undefined
+    /// allocations may be allocated lazily as bytes are initialized but either none or all bytes
+    /// are allocated.
+    crate bytes: Vec<u8>,
```
Can't we restrict this further? Preferrably these would be private to this module only.

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -16,15 +16,20 @@ use std::borrow::Cow;
+    /// Denotes which part of this allocation is initialized
+    crate undef_mask: UndefMask,
+    /// The size of the allocation.
+    /// Used after initialiation of an undef region that is not allocated.
```
```suggestion
    /// Can be bigger than `bytes.len()` because we do not allocate all bytes immediately.
```

I don't understand what the second sentence here is telling me, is this a correct interpretation.

HeroicKatora replied:
That part of the doc shouldn't yet be there, that's part of the changeset for eliding allocation which was intended as a follow up instead :/

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -534,12 +560,90 @@ impl<'tcx, Tag, Extra> Allocation<Tag, Extra> {
         );
     }
+
+    pub fn compress_defined_range(
```
Uh, given the PR title I had hoped this was just refactoring.  Please make sure your title reflects what you did!  Seems like beyond making `byte` private, this also changes the internal representaion?

This is two changes in one. Is there any chance you can split those into separate PRs? That would make them *a lot* easier to review.

HeroicKatora replied:
This just moves the compression from `interpret/memory.rs` to `allocation`. It doesn't (at least shouldn't) change its implementation. It's hard to hide `undef_mask` without moving this since `memory` accesses it extensively. Well, there could be an intermediate PR with `fn undef_mask_mut(&mut self)` but that didn't really seem sensible?

RalfJung replied:
Oh, I missed that this is moved in from there, sorry.

But then why the changes in the `Alloc` struct, the changed doc comment for `bytes` and the new `size` field?

HeroicKatora replied:
Probably missed them when I cherry-picked some of the accessor utilities from the other PR.. Just now realized that they include references to not allocating all of the memory. Sorry for the confusion.

HeroicKatora commented:
Removed `crate` visibility qualifier, only required rather simple changes to the stable hashing. Also removed the portion of comments on `bytes` and `undef_mask` that were not intended to be part of this PR yet.

HeroicKatora commented:
The `size` field makes it much simpler to access the allocation extent as a `Size`, avoiding many additional conversions from the `usize` length of the byte vector. So it might be worth keeping?

RalfJung commented on src/librustc/ich/impls_ty.rs:
```diff
@@ -176,15 +176,20 @@ impl<'a> HashStable<StableHashingContext<'a>> for mir::interpret::Allocation {
+            relocations, align, mutability, size,
             extra: _,
+            .. /* private bytes and undef_mask */
         } = self;
```
Hm, the reason we use a pattern here is to be alerted when new fields are added, so that we don't forget to also hash them.

@oli-obk is there any way to keep the fields private while having reliable stable hashing? Can we derive this, or move the impl to `allocation.rs`?

RalfJung replied:
`#[derive(HashStable)]` exists but I'll leave it to @oli-obk whether we can use that for `Allocation`.

oli-obk replied:
If we implement `HashStable` for `relocations` we should be able to use the derive. Or we just move this impl to `librustc/mir/interpret/allocation.rs`

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -16,15 +16,17 @@ use std::borrow::Cow;
     pub relocations: Relocations<Tag>,
-    /// Denotes undefined memory. Reading from undefined memory is forbidden in miri
-    pub undef_mask: UndefMask,
+    /// Denotes which part of this allocation is initialized
```
```suggestion
    /// Denotes which part of this allocation is initialized.
```

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -16,15 +16,17 @@ use std::borrow::Cow;
-    pub undef_mask: UndefMask,
+    /// Denotes which part of this allocation is initialized
+    undef_mask: UndefMask,
+    /// The size of the allocation.
```
```suggestion
    /// The size of the allocation. Currently, must always equal `bytes.len()`.
```

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -106,13 +109,33 @@ impl<Tag> Allocation<Tag> {
+    }
+
+    /// Look at a slice which may describe undefined bytes or describe a relocation. This differs
+    /// from `get_bytes_with_undef_and_ptr` in that it does no bounds checks and relocation checks
```
`get_bytes_with_undef_and_ptr` also does not do bounds checks.

RalfJung replied:
Also please extend the doc comment to say something like "every use of this method should come with justification for why `get_bytes_with_undef_and_ptr` wasn't used".

RalfJung replied:
Probably it should also say that this should only be used when inspecting the interpreter state "from the outside", not for interpreter execution. Maybe that can be reflected in the name?

HeroicKatora replied:
Maybe `dump_bytes_with_undef_and_ptr` or `uninterpreted_bytes_with_undef_and_ptr`?

RalfJung replied:
`dump` sounds like a debugging thinking dump to stderr.

`uninterpreted` doesn't really convey anything `raw` doesn't convey. `meta_level_inspect_bytes_with_undef_and_ptr` would work for me but I doubt it works for many others here.^^

oli-obk replied:
I'm all for verbose names when that definitely clarifies it. Better than having a method name and misinterpreting what it means because no one reads docs

Centril replied:
> because no one reads docs

@oli-obk  I will have you that I definitely read docs (with a convenient firefox key `rustc`) when doing reviews where I haven't seen a method. =)

HeroicKatora replied:
@RalfJung You're right, those aren't much better. I would find `meta` to be too generic, my first thought would be that it returns a property *of the bytes* rather than returning a raw part of the interpreter state. It works somewhat similar to a debugger attached to the interpreter, doesn't it? Maybe `inspect_internal_raw_bytes_with_undef_and_ptr` makes it clear that this is not to be used as part of interpreter execution.

RalfJung replied:
I was thinking `inspect` as well (not sure `internal` adds anything), but one could reasonably say that the machine also "inspects" its bytes.

`inspect_from_outside` or so... maybe?^^

HeroicKatora replied:
Combining parts of the above: `inspect_bytes_with_undef_and_ptr_outside_interpreter`?

RalfJung replied:
Well, that's really long. But it does convey the crucial point.

HeroicKatora replied:
Maybe the `_bytes`-part could be dropped. 'inspect' and 'with_undef_and_ptr' already make it clear which part of the allocation is returned. That drops it to slightly below half the line length.

RalfJung replied:
So that would be `inspect_with_undef_and_ptr_outside_interpreter`. @oli-obk is that excessive?

oli-obk replied:
No, not excessive, it makes it very clear that something nontrivial is going on and gives a good hint as to what that thing is.

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -534,12 +557,94 @@ impl<'tcx, Tag, Extra> Allocation<Tag, Extra> {
 
+/// Run-length encoding of the undef mask.
+/// Used to copy parts of a mask multiple times to another allocation.
+pub struct AllocationDefinedness {
```
Could you move the two methods relating to this down here, below `struct AllocationDefinedness`, into their own `impl` block?

RalfJung commented on src/librustc/ty/print/pretty.rs:
```diff
@@ -944,10 +944,10 @@ pub trait PrettyPrinter<'tcx>:
                 },
                 (ConstValue::Slice { data, start, end }, ty::Slice(t)) if *t == u8 => {
-                    Some(&data.bytes[start..end])
+                    Some(data.raw_bytes_with_undef_and_ptr(start..end))
```
Please add a comment explaining why using `raw` is okay here.

RalfJung commented on src/librustc/ty/print/pretty.rs:
```diff
@@ -944,10 +944,10 @@ pub trait PrettyPrinter<'tcx>:
                 },
                 (ConstValue::Slice { data, start, end }, ty::Str) => {
-                    let slice = &data.bytes[start..end];
+                    let slice = data.raw_bytes_with_undef_and_ptr(start..end);
```
Please add a comment explaining why using `raw` is okay here.

RalfJung commented on src/librustc_codegen_llvm/consts.rs:
```diff
@@ -437,7 +440,18 @@ impl StaticMethods for CodegenCx<'ll, 'tcx> {
+                        // undef bytes may yield arbitrary values.
+                        //
+                        // FIXME: ignore undef bytes even with representation `!= 0`.
+                        alloc.raw_bytes_with_undef_and_ptr(all_bytes_range)
```
```suggestion
                        //
                        // The `raw` method is okay here because we checked relocations, and because we are doing
                        // this access to inspect the interpreter state (not as part of the interpreter execution).
                        alloc.raw_bytes_with_undef_and_ptr(all_bytes_range)
```

RalfJung commented on src/librustc_codegen_llvm/consts.rs:
```diff
@@ -437,7 +440,18 @@ impl StaticMethods for CodegenCx<'ll, 'tcx> {
                 if self.tcx.sess.target.target.options.is_like_osx {
-                    let sect_name = if alloc.bytes.iter().all(|b| *b == 0) {
+                    let all_bytes_range = 0..alloc.len();
+                    let is_zeroed = alloc.relocations.len() == 0 && {
```
This `alloc.relocations.len() == 0` is new, right? I was told before that this is already inside a section that checks that there are no relocations. So maybe make this an assertion instead?

RalfJung commented on src/librustc_mir/interpret/snapshot.rs:
```diff
@@ -285,12 +286,24 @@ impl<'a, Ctx> Snapshot<'a, Ctx> for &'a Allocation
+            align,
+            mutability,
+            extra: (),
+            ..
```
Same concern with the `..` pattern here.

Maybe add a comment at the `Allocation` type like
```
// NOTE: When adding new fields, make sure to adjust the StableHash impl in ... and the Snapshot impl in ...
```

RalfJung reviewed (commented):
This looks good overall, my notes are mostly editorial / ask for more comments.

Note that I'll be on conference next week and won't have time for reviews.

oli-obk commented on src/librustc_codegen_llvm/consts.rs:
```diff
@@ -456,10 +470,15 @@ impl StaticMethods for CodegenCx<'ll, 'tcx> {
                         section.as_str().as_ptr() as *const _,
                         section.as_str().len() as c_uint,
                     );
+                    if !alloc.relocations.is_empty() {
```
use an `assert` here

197g committed:
Derive HashStable for Allocation

Requires a manual implementation for Relocations since dereferencing to
SortedMap is not always implemented but that one is far more trivial.
Added fields could otherwise be silently forgotten since private fields
make destructing outside the module possible only with `..` pattern
which would then also be applicable to newly added public fields.

HeroicKatora commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -16,15 +16,17 @@ use std::borrow::Cow;
     /// Only the first byte of a pointer is inserted into the map; i.e.,
     /// every entry in this map applies to `pointer_size` consecutive bytes starting
     /// at the given offset.
     pub relocations: Relocations<Tag>,
```
There already exists a method `relocations`, which returns relocations within a range specified by `ptr` and `size`, so I'd name the corresponding accessor `all_relocations` instead.

HeroicKatora replied:
Also, `interpret/memory.rs` tries to access it directly. I'd move code in a similar manner as compressed undef mask.

oli-obk replied:
Sounds good

HeroicKatora replied:
I'm not available until Wednesday next week but the rest should be ready.

oli-obk replied:
take all the time you need

HeroicKatora replied:
The name `fn relocations` is already taken: https://github.com/rust-lang/rust/blob/85ed538d6988c6c82aea8750b306cb793e874294/src/librustc/mir/interpret/allocation.rs#L424-L430.

I'd rename the existing method to `get_relocations` which would be aligned with `get_bytes`, `check_bytes`, `check_relocations` to be concerned with interpreter execution.

RalfJung replied:
Sounds good.
Makes me wonder who will even need the general read-only general accessor. Maybe that should also be `raw_relocations` or so, to make it less appealing to call?

HeroicKatora commented:
Moved the `HashStable` implementation into the module itself, renamed the raw accessor method, added comments for calls to it.

RalfJung commented:
(I'm on a conference for the next week so won't have time for reviews, sorry.)

oli-obk commented on src/librustc_mir/interpret/snapshot.rs:
```diff
@@ -285,12 +286,25 @@ impl<'a, Ctx> Snapshot<'a, Ctx> for &'a Allocation
+        } = self;
+
+        let all_bytes = 0..self.len();
+        // FIXME: why is this access okay?
```
It's OK because we only use it for detecting loops and we extract the undef mask and relocations separately

HeroicKatora replied:
Good to know, included that in the reasoning.

RalfJung commented on src/librustc_codegen_llvm/consts.rs:
```diff
@@ -35,11 +35,21 @@ pub fn const_alloc_to_llvm(cx: &CodegenCx<'ll, '_>, alloc: &Allocation) -> &'ll
         let offset = offset as usize;
         if offset > next_offset {
-            llvals.push(cx.const_bytes(&alloc.bytes[next_offset..offset]));
+            // This `inspect` is okay since we have check that it is not within a relocation, it is
```
```suggestion
            // This `inspect` is okay since we have checked that it is not within a relocation, it is
```

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -106,13 +122,34 @@ impl<Tag> Allocation<Tag> {
+        self.size.bytes() as usize
+    }
+
+    /// Look at a slice which may describe undefined bytes or describe a relocation. This differs
```
```suggestion
    /// Looks at a slice which may describe undefined bytes or describe a relocation. This differs
```

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -106,13 +122,34 @@ impl<Tag> Allocation<Tag> {
+        &self.bytes[range]
+    }
+
+    /// View the undef mask.
```
```suggestion
    /// Returns the undef mask.
```

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -536,6 +573,91 @@ impl<'tcx, Tag, Extra> Allocation<Tag, Extra> {
+/// Transferring the definedness mask to other allocations.
+impl<Tag, Extra> Allocation<Tag, Extra> {
+    /// Creates a run-length encoding of the undef_mask.
+    pub fn compress_defined_range(
```
`compress_undef_range` better reflects that this is about `undef_mask` IMO.

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -536,6 +573,91 @@ impl<'tcx, Tag, Extra> Allocation<Tag, Extra> {
+    }
+
+    /// Apply multiple instances of the run-length encoding to the undef_mask.
+    pub fn mark_compressed_range(
```
What about `mark_compressed_undef_range`?

HeroicKatora replied:
Good point, adding `undef` into the name.

197g committed:
Address naming and comments from reviews

197g committed:
Make allocation relocation field private

197g committed:
Move relocation range copies into allocation

197g committed:
Fixup remaining direct relocation field references

HeroicKatora commented:
The `relocations` field is now also private, the direct access in `interpret/memory.rs` has been moved into the module similar to the previous change for `undef_mask`.

HeroicKatora commented:
Open to feedback on the naming especially since I'm not perfectly content with it either.

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -688,6 +693,56 @@ impl<Tag> DerefMut for Relocations<Tag> {
+/// A partial, owned list of relocations to transfer into another allocation.
+pub struct AllocationRelocations<Tag> {
+    relative_relocations: Vec<(Size, (Tag, AllocId))>,
+}
```
Interesting idea. Long-term (does not have to happen in this PR), it might make sense to combine this with the run-length encoded undef mask into one struct? They are AFAIK only used together, and basically together opaquely represent "metadata" (on top of the raw bytes) that memcpy needs to copy. Then there would be two operations, tentatively named `get_metadata` and `apply_metadata` or so.

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -688,6 +693,56 @@ impl<Tag> DerefMut for Relocations<Tag> {
+                        reloc,
+                        )
+                })
+                );
```
Indentation looks off here.

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -688,6 +693,56 @@ impl<Tag> DerefMut for Relocations<Tag> {
+        size: Size,
+        dest: Pointer<Tag>,
+        length: u64,
+    ) -> AllocationRelocations<Tag> {
```
What about `get_relocations_range` as name?

HeroicKatora replied:
The return value is destination specific since it already offsets correctly. I didn't think `get` would make this sufficiently clear. It would be possible to move the final offsetting with `dest.offset into `mark_relocation_range` but I didn't want to modify it to stay as close to the current code as possible.

RalfJung replied:
Hm, makes sense.

RalfJung reviewed (approved):
LGTM, with the ntis fixed! However, I can't tell whether the arguments made in `codegen_llvm` are correct. Handing over to 

r? @oli-obk

RalfJung commented:
r? @oli-obk 

(Doesn't look like the bots looks into reviews.)

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -536,6 +578,91 @@ impl<'tcx, Tag, Extra> Allocation<Tag, Extra> {
+/// Used to copy parts of a mask multiple times to another allocation.
+pub struct AllocationDefinedness {
+    ranges: smallvec::SmallVec::<[u64; 1]>,
+    first: bool,
```
`inital` might be a better field name? Also please add a doc comment, and maybe reorder the field.

My first reaction when I just saw this "why does it matter if this is the first definedness every created...?"

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -566,6 +693,56 @@ impl<Tag> DerefMut for Relocations<Tag> {
+        }
+    }
+
+    pub fn mark_relocation_range(
```
Please add doc comment that this expects the previous relocations for that range to be already cleared.

197g committed:
Improve documentation around allocation accessors

RalfJung commented on src/librustc/mir/interpret/allocation.rs:
```diff
@@ -581,8 +581,10 @@ impl<'tcx, Tag, Extra> Allocation<Tag, Extra> {
     ranges: smallvec::SmallVec::<[u64; 1]>,
-    first: bool,
+    /// The definedness of the first range.
+    initial: bool,
```
Tiny nit (but I thought of it and now I can't not say it^^): might make more sense to make `initial` the first field.
When someone sees this type, I think it is easier to understand if they see this one first.

Then the doc comment for `ranges` can just say "at the end of each range, the boolean state flips", or something like that.

HeroicKatora replied:
I see the point, now the doc of `ranges` references `initial` and ready well in natural sequence order.

197g committed:
Reorder AllocationDefinedness members

This improves the clarity of the documentation a bit since they can
reference each other when reading the member docs in sequence.

HeroicKatora commented:
@oli-obk I fixed the remaining nits from Ralf now, there are some unresolved review comments where you commented and I wasn't certain they were fully addressed. And the review of codegen is still oustanding as mentioned. Status should now be waiting on review.

oli-obk commented:
@bors r+

Oh, right. I reviewed those comments when you made them. Reviewed again, still all good.

bors commented:
:pushpin: Commit f3c435eb780dec5c1d390a3b43cdfb4f1528d70d has been approved by `oli-obk`

<!-- @bors r=oli-obk f3c435eb780dec5c1d390a3b43cdfb4f1528d70d -->
<!-- homu: {"type":"Approved","sha":"f3c435eb780dec5c1d390a3b43cdfb4f1528d70d","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit f3c435eb780dec5c1d390a3b43cdfb4f1528d70d with merge b50520835f3b14409797139eca23f401a529e711...
<!-- homu: {"type":"BuildStarted","head_sha":"f3c435eb780dec5c1d390a3b43cdfb4f1528d70d","merge_sha":"b50520835f3b14409797139eca23f401a529e711"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=7227)
Approved by: oli-obk
Pushing b50520835f3b14409797139eca23f401a529e711 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=7227"},"merge_sha":"b50520835f3b14409797139eca23f401a529e711"} -->

rust-highfive commented:
📣 Toolstate changed by rust-lang/rust#63561!

Tested on commit rust-lang/rust@b50520835f3b14409797139eca23f401a529e711.
Direct link to PR: <https://github.com/rust-lang/rust/pull/63561>

💔 clippy-driver on windows: test-pass → build-fail (cc @Manishearth @llogiq @mcarton @oli-obk @phansch @flip1995, @rust-lang/infra).
💔 clippy-driver on linux: test-pass → build-fail (cc @Manishearth @llogiq @mcarton @oli-obk @phansch @flip1995, @rust-lang/infra).
💔 miri on windows: test-pass → build-fail (cc @oli-obk @RalfJung @eddyb, @rust-lang/infra).
💔 miri on linux: test-pass → build-fail (cc @oli-obk @RalfJung @eddyb, @rust-lang/infra).
